File: src/multi/algo/BDwithPAP_Reactive.java
package multi.algo;

import ilog.concert.IloException;
import ilog.cplex.IloCplex;
import multi.*;
import multi.model.DetermineModel;
import multi.model.DualSubProblem;
import multi.model.DualSubProblemReactive;
import multi.model.MasterProblem;
import multi.data.InputData;
import multi.data.Parameter;
import multi.data.Scenario;
import java.io.IOException;
import java.util.List;

public class BDwithPAP_Reactive extends BD {

    private DualSubProblemReactive dsp;
    public BDwithPAP_Reactive(InputData in, Parameter p) throws IloException, IOException {
        super();
        this.in = in;
        this.p = p;
        this.tau = p.getTau();
        this.Algo = "BD&PAP-Reactive";
        this.AlgoID = Algo + "-R"+ in.getShipRouteSet().size() + "-T" + p.getTimeHorizon() + "-"+ FleetType + "-S" + randomSeed + "-V" + VesselCapacityRange;
        frame();
    }
    public BDwithPAP_Reactive(InputData in, Parameter p, int tau) throws IloException, IOException {
        super();
        this.in = in;
        this.p = p;
        this.tau = tau;
        this.Algo = "BD&PAP-Reactive";
        this.AlgoID = Algo + "-R"+ in.getShipRouteSet().size()
                + "-T" + p.getTimeHorizon()
                + "-"+ FleetType
                + "-S" + randomSeed
                + "-V" + VesselCapacityRange;
        frame();
    }
    @Override
    protected double initialModel() throws IloException, IOException {
        double start = System.currentTimeMillis();

        dsp =new DualSubProblemReactive(in, p, tau);
        mp=new MasterProblem(in, p, "Reactive");

        if(WhetherAddInitializeSce){
            mp.addScene(sce.get(0));
        }

        if(WhetherSetInitialSolution){
            DetermineModel dm = new DetermineModel(in, p);
            mp.setInitialSolution(dm.getVVarValue());
        }

        return System.currentTimeMillis() - start;
    }

    @Override
    protected void frame() throws IloException, IOException {
        initialize();

        if(WhetherAddInitializeSce)
            initializeSce(sce);

        // change  MaxVarDemand
        // beta = min(k, m/k)=tau

        double[] maxDemandVar = p.getMaximumDemandVariation();
        p.changeMaximunDemandVariation(tau);

        double time0 = System.currentTimeMillis();
        initialModel();

        printIterTitle(fileWriter, System.currentTimeMillis() - time0);
        printIteration(fileWriter, lower[iteration], upper[iteration],0, 0, 0,
                "--", 0, "--", 0 );

        // add the initial scene to make the MP feasible
        /*initializeSce(sce);*/
        int flag = 0;
        mp.addScene(sce.get(iteration));
        double start0 = System.currentTimeMillis();
        while(upperBound - lowerBound > boundGapLimit
                && flag == 0
                && iteration<maxIterationNum
                && (System.currentTimeMillis() - start0)/1000 < maxIterationTime
        )
        {
            double start1 = System.currentTimeMillis();
            mp.solveReactiveModel();
            double end1 = System.currentTimeMillis();

            // check if the mp-solution changed
            if(! addSolutionPool(mp.getSolution())){
                System.out.println("MP solution duplicate");
                flag=1;
            }

            // LB = max{LB , MP.Objective}
            // LB = MP.Objective = MP.OperationCost + Eta
            if(mp.getObjVal()>lowerBound
                    && mp.getSolveStatus() == IloCplex.Status.Optimal) {
                setLowerBound(mp.getObjVal());
            }

            dsp.changeObjectiveVCoefficients(mp.getVVarValue(), mp.getVVarValue2());
            double start2 = System.currentTimeMillis();
            dsp.solveModel();
            double end2 = System.currentTimeMillis();

            //  the SP is optimal :  add optimality cut
            if(dsp.getSolveStatus() == IloCplex.Status.Optimal )
            {
                //  update UB : UB = min{UB, MP.OperationCost + SP.Objective}
                if(dsp.getObjVal()+mp.getOperationCost()<upperBound)
                {
                    setUpperBound(dsp.getObjVal()+mp.getOperationCost());
                }

                // add optimality cut
                mp.addReactiveOptimalityCut(dsp.getConstantItem(), dsp.getBeta1Value(), dsp.getBeta2Value());

                // add worse case to scene set
                if(! addScenarioPool(dsp.getScene())){
                    sce.add(dsp.getScene());
                }
            }
            // the SP is unbounded : add feasibility cut
            else if(dsp.getSolveStatus() == IloCplex.Status.Unbounded)
            {
                // ! here beta is extreme ray !
                mp.addReactiveFeasibilityCut(dsp.getConstantItem(), dsp.getBeta1Value(), dsp.getBeta2Value());
            }

            iteration++;
            upper[iteration]=upperBound;
            lower[iteration]=lowerBound;

            if(WhetherPrintProcess || WhetherPrintIteration){
                System.out.println(iteration+"\t\t"
                        +String.format("%.2f", upper[iteration])+"\t\t"
                        +String.format("%.2f", lower[iteration])+"\t\t"
                        +String.format("%.2f", end2 - start2)+"\t\t"
                        +String.format("%.2f", end1 - start1)+"\t\t"
                        +String.format("%.2f",System.currentTimeMillis() - start0)+"\t\t"
                        +dsp.getSolveStatusString()+"("+String.format("%.4f", dsp.getMipGap())+")"+"\t\t"
                        +mp.getSolveStatusString()+"("+String.format("%.4f", mp.getObjGap())+")"
                );
            }
            if(DefaultSetting.WhetherWriteFileLog){
                fileWriter.write(iteration+"\t\t"
                        +String.format("%.2f", upper[iteration]) +"\t\t"
                        +String.format("%.2f", lower[iteration])+"\t\t"
                        +String.format("%.2f", end2 - start2)+"\t\t"
                        +String.format("%.2f", end1 - start1)+"\t\t"
                        +String.format("%.2f",System.currentTimeMillis() - start0)+"\t\t"
                        +dsp.getSolveStatusString()+"("+String.format("%.4f", dsp.getMipGap())+")"+"\t\t"
                        +mp.getSolveStatusString()+"("+String.format("%.4f", mp.getObjGap())+")"
                );
                fileWriter.write("\n");
            }
        }

        setAlgoResult();
        end();

        p.setMaximumDemandVariation(maxDemandVar);
    }

    @Override
    protected void initializeSce(List<Scenario> sce)
    {
        double[] sss  =new double [in.getRequestSet().size()];

        // |v| = tau/I |e| = tau/I
        double v=(double)tau/(double)p.getDemand().length * 1/ (Math.sqrt(p.getDemand().length));
        for(int i=0;i<p.getDemand().length;i++)
        {
            sss[i]=v ;
        }

        sce.add(new Scenario(sss));
    }
}

File: Multi_routes_deploy/sources/algo/BDwithPAP_Reactive.cpp
#include "BDwithPAP_Reactive.h"

namespace fleetdeployment
{

    void BDwithPAP_Reactive::frame() {
        std::string filename = "BD&PAP&Reactive" + std::to_string(p_.GetTimePointSet().size()) + ".txt";
        std::ofstream fileWriter(filename, std::ios::app);

        if (!fileWriter.is_open()) {
            std::cerr << "Error: Unable to open file " << filename << std::endl;
        }

        time_t start = time(0);

        // change  MaxVarDemand
        // beta = min(k, m/k)=tau

        SetUpperBound(1e12);
        SetLowerBound(-1e12);

        std::vector<Scenario> sce;

        std::vector<double> maxDemandVar = p_.GetMaximumDemandVariation();
        p_.changeMaximunDemandVariation(tau);

        int max_iteration = 100;
        std::vector<std::vector<int>> vvv(p_.GetVesselSet().size(), std::vector<int>(p_.GetVesselRouteSet().size()));
        std::vector<double> upper(max_iteration + 1);
        std::vector<double> lower(max_iteration + 1);
        upper[0] = upperBound;
        lower[0] = lowerBound;
        std::vector<int> worseCaseSet(p_.GetDemand().size());
        double gap = std::numeric_limits<double>::max();

        time_t time0 = time(0);

        MasterProblem mp(in_, p_, true);
        DualSubProblemReactive dsp(in_, p_, 1);

        // add the initial scene to make the MP feasible
        Initialize(sce);
        int iteration = 0;
        int flag = 0;

        if (WhetherPrintProcess) {
            std::cout << "==========BD with PAP Reactive: ===========" << std::endl;
            std::cout << "BuildModelTime = "
                << std::fixed << std::setprecision(2) << time(0) - time0 << std::endl;
            std::cout << "k" << "\t\t"
                << "UB" << "\t\t"
                << "LB" << "\t\t"
                << "DSP-SolveTime" << "\t\t"
                << "MP-SolveTime" << "\t\t"
                << "Total Time" << "\t\t"
                << "DSP-Status" << "\t\t"
                << "MP-Status" << std::endl;
            std::cout << iteration << "\t\t"
                << std::fixed << std::setprecision(2) << upper[iteration] << "\t\t"
                << std::fixed << std::setprecision(2) << lower[iteration] << std::endl;
        }
        if (WhetherOutputLog) {
            fileWriter << "k" << "\t\t"
                << "UB" << "\t\t"
                << "LB" << "\t\t"
                << "DSP-SolveTime" << "\t\t"
                << "MP-SolveTime" << "\t\t"
                << "Total Time" << "\t\t"
                << "DSP-Status" << "\t\t"
                << "MP-Status" << std::endl;
            fileWriter << iteration << "\t\t"
                << std::fixed << std::setprecision(2) << upper[iteration] << "\t\t\t"
                << std::fixed << std::setprecision(2) << lower[iteration] << std::endl;
        }

        mp.addScene(sce[iteration]);
        time_t start0 = time(0);
        while (upperBound - lowerBound > MIPGapLimit
            && gap > MIPGapLimit
            && flag == 0
            && iteration < max_iteration)
        {
            time_t start1 = time(0);
            mp.SolveReactiveModel();
            time_t end1 = time(0);

            // check if the mp-solution changed
            double temp = 0;
            for (int r = 0; r < p_.GetVesselRouteSet().size(); r++)
            {
                for (int h = 0; h < p_.GetVesselSet().size(); h++)
                {
                    temp = temp + abs(mp.GetVVarValue()[h][r] - vvv[h][r]);
                    vvv[h][r] = mp.GetVVarValue()[h][r];
                }
            }
            // master solution has no change
            if (temp < 0.5)
            {
                flag = 1;
            }

            // LB = max{LB , MP.Objective}
            // LB = MP.Objective = MP.OperationCost + Eta
            if (mp.GetObjVal() > lowerBound) {
                SetLowerBound(mp.GetObjVal());
            }

            dsp.changeObjectiveVCoefficients(mp.GetVVarValue(), mp.GetVVarValue2());
            time_t start2 = time(0);
            dsp.SolveModel();
            time_t end2 = time(0);

            //  the SP is optimal :  add optimality cut
            if (dsp.GetSolveStatus() == IloCplex::Status::Optimal)
            {
                //  update UB : UB = min{UB, MP.OperationCost + SP.Objective}
                if (dsp.GetObjVal() + mp.GetOperationCost() < upperBound)
                {
                    SetUpperBound(dsp.GetObjVal() + mp.GetOperationCost());
                }

                // add optimality cut
                mp.addReactiveOptimalityCut(dsp.GetConstantItem(), dsp.GetBeta1Value(), dsp.GetBeta2Value());
            }
            // the SP is unbounded : add feasibility cut
            else if (dsp.GetSolveStatus() == IloCplex::Status::Unbounded)
            {
                // ! here beta is extreme ray !
                mp.addReactiveFeasibilityCut(dsp.GetConstantItem(), dsp.GetBeta1Value(), dsp.GetBeta2Value());
            }

            iteration = iteration + 1;
            upper[iteration] = upperBound;
            lower[iteration] = lowerBound;
            gap = upper[iteration] - lower[iteration];

            if (WhetherPrintProcess || WhetherPrintIteration) {
                std::cout << iteration << "\t\t"
                    << std::fixed << std::setprecision(2) << upper[iteration] << "\t\t"
                    << std::fixed << std::setprecision(2) << lower[iteration] << "\t\t"
                    << std::fixed << std::setprecision(2) << difftime(end2, start2) << "\t\t"
                    << std::fixed << std::setprecision(2) << difftime(end1, start1) << "\t\t"
                    << std::fixed << std::setprecision(2) << difftime(time(0), start0) << "\t\t"
                    << dsp.GetSolveStatusString() << "(" << std::fixed << std::setprecision(4) << dsp.GetMipGap() << ")\t\t"
                    << mp.GetSolveStatusString() << "(" << std::fixed << std::setprecision(4) << mp.GetMipGap() << ")\n";
            }
            if (WhetherOutputLog) {
                fileWriter << iteration << "\t\t"
                    << std::fixed << std::setprecision(2) << upper[iteration] << "\t\t"
                    << std::fixed << std::setprecision(2) << lower[iteration] << "\t\t"
                    << std::fixed << std::setprecision(2) << difftime(end2, start2) << "\t\t"
                    << std::fixed << std::setprecision(2) << difftime(end1, start1) << "\t\t"
                    << std::fixed << std::setprecision(2) << difftime(time(0), start0) << "\t\t"
                    << dsp.GetSolveStatusString() << "(" << std::fixed << std::setprecision(4) << dsp.GetMipGap() << ")\t\t"
                    << mp.GetSolveStatusString() << "(" << std::fixed << std::setprecision(4) << mp.GetMipGap() << ")\n";
            }
        }

        mp.end();
        dsp.end();

        SetVValue(mp.GetVVarValue());
        SetIter(iteration);
        SetObj(upperBound);

        Gap = (upperBound - lowerBound) / lowerBound;

        if (WhetherPrintProcess) {
            PrintSolution();
        }

        time_t end = time(0);
        SetSolveTime(difftime(end, start));

        if (WhetherPrintProcess || WhetherPrintSolveTime) {
            std::cout << "BD SolveTime = " << (difftime(end, start)) << std::endl;
            std::cout << "=================================" << std::endl;
        }
        if (WhetherOutputLog) {
            fileWriter << "BD SolveTime = " << difftime(end, start) << std::endl;
        }
        fileWriter.close();

        p_.SetMaximumDemandVariation(maxDemandVar);
    }


    void BDwithPAP_Reactive::Initialize(std::vector<Scenario>& sce)
    {
        std::vector<double> sss(in_.GetRequests().size());

        // beta = min{k , I/k}
        double beta = (double)tau > p_.GetDemand().size() / (double)tau ?
            p_.GetDemand().size() / (double)tau : (double)tau;

        double v = (double)1 / (double)p_.GetDemand().size();
        for (int i = 0; i < p_.GetDemand().size(); i++)
        {
            sss[i] = beta * v * 1 / sqrt(p_.GetDemand().size());
        }

        Scenario sp;
        sp.SetRequest(sss);
        sce.push_back(sp);
    }


    void BDwithPAP_Reactive::PrintSolution() {
        std::cout << "Vessel Decision vVar : ";
        for (int r = 0; r < p_.GetVesselRouteSet().size(); r++)
        {
            std::cout << p_.GetVesselRouteSet()[r] << ":\t";
            for (int h = 0; h < p_.GetVesselSet().size(); h++)
            {
                if (vValue[h][r] != 0)
                {
                    std::cout << "(" << p_.GetVesselSet()[h] << ")\t";
                }
            }
        }
        std::cout << std::endl;
    }



}